proposals/NNNN-package-access-modifier.md

```diff
@@ -0,0 +1,282 @@
+
+## Detailed design
+
+### `package` Keyword
+
+`package` is introduced as an access modifier.  It cannot be combined with other access modifiers.
+`package` is a contextual keyword, so existing declarations named `package` will continue to work.  This follows the precedent of `open`, which was also added as a contextual keyword.  For example, the following is allowed:
+
+```
+package var package: String {...}
+```
+
+This is similar to `open`, which is also a contextual keyword.
+
+### Declaration Site
+
+The `package` keyword is added at the declaration site.  Using the scenario above, the helper API `run` can be declared with the new access modifier like so:
+
+```
+[Engine]
+
+public struct MainEngine {
+    public init() { ...  }
+    public var stats: String { ...  }
+    package func run() { ...  }
+}
+```
+
+The `package` access modifier can be added to any types where an existing access modifier can be added, e.g. `class`, `struct`, `enum`, `func`, `var`, `protocol`, etc.  It is less restrictive than `internal` and more restrictive than `public`.  For example, a `public` function cannot have `internal` or `package` parameters or return type in its signature, and a `package` function cannot have `internal` parameters or return type in its signature.  
+
+
+### Use Site
+
+The Game module can access the helper API `run` since it is in the same package as Engine.
+
+```
+[Game]
+
+import Engine
+
+public func play() {
+    MainEngine().run() // Can access `run` as it is a package symbol in the same package
+}
+```
+
+However, if a client outside of the package tries to access the helper API, it will not be allowed.
+
+```
+[App]
+
+import Game
+import Engine
+
+let engine = MainEngine()
+engine.run() // Error: cannot find `run` in scope
+
+```
+### Package Names
+
+A package name is a unique string with c99 identifier characters, and is passed down to Swift frontend via a new flag `-package-name`.  It is then stored in the module binary and used to compare with package names of other modules to determine if they are part of the same package.  
+
+If `-package-name` is not given, the `package` access modifier is disallowed.  Swift code that does not use `package` access will continue to build without needing to pass in `-package-name`.
+
+Here's an example of how a package name is passed to a commandline invocation.
+
+```
+swiftc -module-name Engine -package-name gamePkg ...
+[Game] swiftc -module-name Game -package-name gamePkg ...
+[App] swiftc App -package-name appPkg ...
+```
+
+When building the Engine module, the package name 'gamePkg' and package symbols is stored in the Engine binary.  When building Game, the input to its package name 'gamePkg' is compared with Engine's package name; since they match, access to package symbols is allowed.  When building App, the name comparison shows 'appPkg' is different from its dependency module's so access to package symbols is deined, which is what we want.
```

Comment:
The "binary" is not a language concept.  (It's also not a precise concept: the compiled code is a "binary", but the package name isn't part of that; also, we may want to record this in `swiftinterface`s eventually.)  Suggestion:

```suggestion
When building the Engine module, the package name 'gamePkg' is recorded in the built interface to the module.  When building Game, its package name 'gamePkg' is compared with the package name recorded in Engine's built interface; since they match, Game is allowed to access Engine's `package` declarations.  When building App, its package name 'appPkg' is different from `gamePkg`, so it is not allowed to access `package` symbols in either Engine or Game, which is what we want.
```